### ISSUE TITLE
Multiple Initial Guess Modes for InverseKinematics

### ISSUE THREAD

**Issue by Ipuch** [Ipuch/bioNC#101]
**Current Behavior**:
The default behavior of the InverseKinematics class aims to compute initial guesses for joint angles using marker data. The approach relies on a function defined by the model_creator from the c3d module.

This function is invoked internally using the method model.Q_from_markers, which approximates joint angles Q based on the marker positions. An optimization process further refines these initial guesses to ensure that they adhere to specified constraints.

``` python3
self.model.Q_from_markers(self.experimental_markers[:, :, :])
```

For users, the interface is kept simple and does not require any manual initialization for the initial guesses:

``` python3
ik_solver = InverseKinematics(
        model, markers, solve_frame_per_frame=True, 
        active_direct_frame_constraints=False, use_sx=True, Q_init=None
    )
Qopt_ipopt = ik_solver.solve(method="ipopt")
```

**Proposed Enhancement**:
To provide more flexibility, we can introduce multiple modes for initial guesses. I propose the following initial guess modes:

- Experimental:
Usage: InverseKinematics(..., initial_guess_mode=InitialGuessMode.FROM_CURRENT_MARKERS)
Description: This is the default behavior. It uses the experimental marker data to form initial guesses. No Q_init input is needed.
- Experimental First Frame:
Usage: InverseKinematics(..., initial_guess_mode=InitialGuessMode.FROM_FIRST_FRAME_MARKERS)
Description: This mode takes the experimental data from the first frame as the initial guess. This guess is then used as a starting point or "warm start" for subsequent frame optimizations. This approach is iterative, refining the guess with each frame until the entire sequence is processed. No Q_init is required here either.
- Custom:
Usage: InverseKinematics(..., initial_guess_mode=InitialGuessMode.USER_PROVIDED)
Description: In this mode, users can provide their custom initial guesses for joint angles with the Q_init argument.

Here is the Enum to implement:

``` python3
class InitialGuessMode(Enum):
    FROM_CURRENT_MARKERS = "From Current Markers"
    FROM_FIRST_FRAME_MARKERS = "From First Frame Markers"
    USER_PROVIDED = "User Provided"
```

Referring to the email discussion we had this summer @anmuller :)

**Comment by aaiaueil** [comment]
To be sure I understand correctly: in Experimental First Frame mode, the experimental first frame is used as initial guess for all frames, or just for the first one? If it's just for the first one, how would the initial guesses for all the other frames be computed?

**Comment by Ipuch** [comment]
The first initial guess is experimental, the next initial guess is the result of the first optimization. The third initial guess is the result of the second optimization etc...

**Comment by Ipuch** [comment]
To precise my thoughts, here how the interface should look like after the PR:
```python3
ik_solver = InverseKinematics(
        model, 
        markers, 
        solve_frame_per_frame=True, 
        active_direct_frame_constraints=True, 
        use_sx=True, 
        #Q_init=None # this should be displaced in `solve` method !!
    )
```
First case, Q_init is not needed, not possible yet with heatmaps:
```python3   
Qopt_ipopt = ik_solver.solve(
        method="ipopt",
        initial_guess_mode=InitialGuessMode.FROM_CURRENT_MARKERS,
)
```

Second case, Q_init is provided for every frames:
```python3   
Qopt_ipopt = ik_solver.solve(
        method="ipopt",
        initial_guess_mode=InitialGuessMode.USER_PROVIDED,
        Q_init=Q_init, #  [nb_q x nb_segment x nb_frames]
)
```

Third case, Q_init is provided for the first frame only:
```python3   
Qopt_ipopt = ik_solver.solve(
        method="ipopt",
        initial_guess_mode=InitialGuessMode.USER_PROVIDED_FIRST_FRAME_ONLY,
        Q_init=Q_init,   # it means that nb_frame = 1 in this case its a [nb_q x nb_segment]
)
```

Fourth case, Q_init is estimated from markers from the first frame only, not possible yet with heatmaps:
```python3   
Qopt_ipopt = ik_solver.solve(
        method="ipopt",
        initial_guess_mode=InitialGuessMode.FROM_FIRST_FRAME_MARKERS,
)
```

**Comment by aaiaueil** [comment]
Just to be sure I get it correctly, should InitialGuessMode be a class? Or what kind of object should it be? From what I gather it does not look like a if...elif loop is the kind of structure you're expecting when coding the initial guess mode type

**Comment by Ipuch** [comment]
Yes, it's a class, to be more precise it's an Enum class, see up there my previous post.
Yes, we have no choice here we will end with a "if else somewhere", but we need to be sure it will append only appear. If you start writing "if else" everywhere because of these changes, then, it might not be the right spot.

**Comment by aaiaueil** [comment]
Hey, we were thinking with @ANaaim of something like this :

``````python

import numpy as np
from enum import Enum
       

class InitialGuessModeBase(ABC):

    def __init__(self,var1:str,var2,var3):
        self.var1 = var1
        self.toto = var2*var3

    @abstractmethod
    def method1(self, alpha:int,beta:int,gamma:int):
    
    @abstractmethod
    def method2(self, titi, toto, tutu):


class InitialGuessMode:

    class FromCurrentMarkers(InitialGuessModeBase):

    class UserProvided(InitialGuessModeBase):

    class UserProvidedFirstFrameOnly(InitialGuessModeBase):

    class FromFirstFrameMarkers(InitialGuessModeBase):


class InitialGuessModeType(Enum):
    FROM_CURRENT_MARKERS = InitialGuessMode.FromCurrentMarkers,
    USER_PROVIDED = InitialGuessMode.UserProvided,
    USER_PROVIDED_FIRST_FRAME_ONLY = InitialGuessMode.UserProvidedFirstFrameOnly,
    FROM_FIRST_FRAME_MARKERS = InitialGuessMode.FromFirstFrameMarkers
``````

And in enums.py adding something like :

``````python
class InitialGuessModeType(Enum):

    TOTO = 1
    TITI = 2
    TUTU = 3
``````

What do you think of this architecture?

**Comment by Ipuch** [comment]
I don't understand why so much complexity is required here. My advice would be implement something that works, and we will see what's going through the PR review. Maybe I pushed you in the wrong direction by giving you too much implementation details, while I seemed very clear to me the simplicity.

Sometimes, just don't be afraid of doing something that works first and iterate yourself the second day. This should be very easy to make it work at first.
